Src/Support/GoogleApis/Apis/[Media]/Download/MediaDownloader.cs

```diff
@@ -166,96 +175,86 @@ public IDownloadProgress Download(string url, Stream stream)
-                        long mediaContentLength;
-
-                        if (contentRange == null)
-                        {
-                            // Content range is null when the server doesn't adhere the media download protocol, in 
-                            // that case we got all the media in one chunk.
-                            currentRequestFirstBytePos = mediaContentLength =
-                                response.Content.Headers.ContentLength.Value;
-                        }
-                        else
-                        {
-                            currentRequestFirstBytePos = contentRange.To.Value + 1;
-                            mediaContentLength = contentRange.Length.Value;
-                        }
+                        throw await MediaApiErrorHandling.ExceptionForResponseAsync(service, response).ConfigureAwait(false);
+                    }
 
-                        if (currentRequestFirstBytePos == mediaContentLength)
+                    // Read up to ChunkSize bytes at a time from the response stream. Write each
+                    // chunk to the caller's stream and notify the caller of progress.
+                    using (var responseStream = await response.Content.ReadAsStreamAsync().ConfigureAwait(false))
+                    {
+                        while (true)
                         {
-                            var progress = new DownloadProgress(DownloadStatus.Completed, mediaContentLength);
-                            UpdateProgress(progress);
-                            return progress;
+                            // Read up to buffer length (ChunkSize) or until end-of-stream.
+                            // ReadAsync will return if it has *any* data available, so we loop.
+                            int bufferOffset = 0;
+                            while (bufferOffset < buffer.Length)
+                            {
+                                int read = await responseStream.ReadAsync(buffer, bufferOffset, buffer.Length - bufferOffset, cancellationToken).ConfigureAwait(false);
+                                if (read == 0)
+                                {
+                                    break;
+                                }
+
+                                // This can't overflow since it's bounded by ChunkSize.
+                                bufferOffset += read;
+                            }
+
+                            await stream.WriteAsync(buffer, 0, bufferOffset, cancellationToken).ConfigureAwait(false);
+                            checked { bytesRead += bufferOffset; }
+
+                            // If we read less than ChunkSize, we've reached the end of the stream.
+                            // In that case, avoid reporting a Downloading event since we'll instead
+                            // report Completed.
+                            if (bufferOffset < buffer.Length)

```

Comment:
Hm, sure enough. That doesn't match the previous behavior. I'll look for a way to avoid this without mangling the code too much.